csharp/src/IceLocatorDiscovery/PluginI.cs

```diff
@@ -256,350 +173,224 @@ public void
+                    _findLocatorTask = null;
                 }
                 else
-                {
-                    l = _locator;
-                }
-
-                if (l != null)
                 {
                     // We found another locator replica, append its endpoints to the current locator proxy endpoints,
                     // while eliminating duplicates.
-                    var newEndpoints = l.Endpoints.Concat(locator.Endpoints).Distinct();
-                    l = l.Clone(endpoints: newEndpoints);
-                }
-                else
-                {
-                    l = locator;
-                }
-
-                if (_pendingRequests.Count == 0)
-                {
-                    _locators[locator.Identity.Category] = l;
-                    Monitor.Pulse(this);
-                }
-                else
-                {
-                    _locator = l;
-                    if (_instanceName.Length == 0)
-                    {
-                        _instanceName = _locator.Identity.Category; // Stick to the first locator
-                    }
-
-                    //
-                    // Send pending requests if any.
-                    //
-                    foreach (Request req in _pendingRequests)
-                    {
-                        _ = req.Invoke(_locator);
-                    }
-                    _pendingRequests.Clear();
+                    _locator = _locator.Clone(endpoints: _locator.Endpoints.Concat(locator.Endpoints).Distinct());
                 }
             }
         }
 
-        public void
-        Invoke(ILocatorPrx? locator, Request? request)
+        private async Task<ILocatorPrx> GetLocatorAsync()
         {
-            lock (this)
+            Task<ILocatorPrx> findLocatorTask;
+            lock (_mutex)
             {
-                if (request != null && _locator != null && _locator != locator)
+                // If we already have a locator we use it.
+                if (_locator != null)
                 {
-                    _ = request.Invoke(_locator);
+                    return _locator;
                 }
-                else if (request != null && Time.CurrentMonotonicTimeMillis() < _nextRetry)
+                // If the retry delay has not elapsed since the last failure return the void locator that always
+                // replies with a null proxy.
+                else if (Time.CurrentMonotonicTimeMillis() < _nextRetry)
                 {
-                    _ = request.Invoke(_voidLocator); // Don't retry to find a locator before the retry delay expires
+                    return _voidLocator;
                 }
+                // If a locator lookup is running we await on it otherwise we start a new lookup.
                 else
                 {
-                    _locator = null;
-
-                    if (request != null)
+                    if (_findLocatorTask == null)
```

Comment:
We need to check if the returned task is completed or not in case it completes synchronously.

```suggestion
                    if (_findLocatorTask == null)
                    {
                        findLocatorTask = FindLocatorAsync();
                        if(!findLocatorTask.IsCompleted)
                        {
                             _findLocatorTask = findLocatorTask;
                        }
                    }
```

Reply:
I don't understand your suggestion, with this change `findLocatorTask ` will be only initialized when `_findLocatorTask` is null, why does it matter if it completes synchronously here?

Reply:
What about

```
                // If a locator lookup is running we await on it otherwise we start a new lookup.
                else
                {
                    if (_findLocatorTask == null)
                    {
                        _findLocatorTask = FindLocatorAsync();
                    }
                    findLocatorTask = _findLocatorTask;
                }
            }
            ILocatorPrx locator = await findLocatorTask.ConfigureAwait(false);
            lock (_mutex)
            {
                _findLocatorTask = null;
            }
            return locator;
```